implementations/rust/ockam/ockam_api/src/authenticator/direct.rs

```diff
@@ -0,0 +1,419 @@
+                        Response::ok(req.id()).body(member).to_vec()?
+                    } else {
+                        Response::not_found(req.id()).to_vec()?
+                    }
+                }
+                // Validate member credential.
+                ["credential"] => {
+                    let crd: Credential = dec.decode()?;
+                    if let Ok(att) = self.signer.verify(&crd).await {
+                        if let Some(err) = check_credential(&self.m_store, &req, &att).await? {
+                            err.to_vec()?
+                        } else {
+                            Response::ok(req.id())
+                                .body(Cbor(crd.attributes()))
+                                .to_vec()?
+                        }
+                    } else {
+                        Response::unauthorized(req.id()).to_vec()?
+                    }
+                }
+                _ => response::unknown_path(&req).to_vec()?,
+            },
+            Some(Method::Delete) => match req.path_segments::<3>().as_slice() {
+                // Admin wants to remove an enroller.
+                ["enroller", id] if self.is_admin(from) => {
+                    self.e_store.del(ENROLLER, id).await?;
+                    Response::ok(req.id()).to_vec()?
+                }
+                ["enroller", _] => Response::unauthorized(req.id()).to_vec()?,
+                _ => response::unknown_path(&req).to_vec()?,
+            },
+            _ => response::invalid_method(&req).to_vec()?,
+        };
+
+        Ok(res)
+    }
+
+    /// Check that the given identity ID matches the one configured as admin.
+    fn is_admin(&self, k: &IdentityIdentifier) -> bool {
+        Some(k.key_id().as_str()) == self.admin.as_ref().map(|a| a.as_str())
+    }
+}
+
+async fn check_credential<'a, S: AuthenticatedStorage>(
+    store: &S,
+    req: &'a Request<'_>,
+    attrs: &Attributes<'_>,
+) -> Result<Option<ResponseBuilder<Error<'a>>>> {
+    // Verify that the credential has not expired
+
+    let ts = if let Some(val) = attrs.get("ts") {
+        minicbor::decode(val)?
+    } else {
+        return Ok(Some(response::bad_request(
+            req,
+            "missing credential timestamp",
+        )));
+    };
+    let now = Timestamp::now().ok_or_else(invalid_sys_time)?;
+    if let Some(dur) = now.elapsed(ts) {
+        if dur > MAX_VALIDITY {
+            return Ok(Some(response::forbidden(req, "credential expired")));
+        }
+    } else {
+        return Ok(Some(response::bad_request(
+            req,
+            "invalid credential timestamp",
+        )));
+    }
+
+    // Verify that the member is authorised
+
+    let id = if let Some(val) = attrs.get("id") {
+        str::from_utf8(val).map_err(invalid_utf8)?
+    } else {
+        return Ok(Some(response::bad_request(
+            req,
+            "missing credential identity",
+        )));
+    };
+    check_member(store, req, id).await
```

Comment:
I don't get the following fact: if we can verify that presenter is a member according to our storage, why would we need a credential? Or if have a valid credential presented, why would we check membership in our storage?

Reply:
When the authenticator runs in the cloud as part of the project's authority and some entity asks for the credential validation it may be that despite a valid credential the enroller has previously removed the member (not implemented yet), hence the validation should fail. I am not sure how useful the authenticator is on client side where we do not issue credentials.

Reply:
This is related to the discussion below - about separation of the authenticator and verifiers. Authenticator is mostly dedicated to issuing credentials, while verifier doesn't have access to the members storage. Is that right?

Reply:
It depends on the kind of verification that is be done. The signer verifies the signature only and leaves it to other parts of the system to inspect the credential further. If this authenticator is used to verify a credential (I imagined this would perhaps be used in the sidecar scenario) it considers its member storage.

Reply:
Pure client-side credential verification is not implemented beyond using the signer to verify some credential. I think this would be part of https://github.com/build-trust/codebase/issues/369.